### DISCUSSION TITLE
Supporting custom store types

### DISCUSSION THREAD

**Discussion by liammacisaac** [ebourg/jsign#202] [General]
As discussed in https://github.com/ebourg/jsign/discussions/199, I'm implementing a custom `SigningService` and have done so by adding to the `KeyStoreType` enum for now. However, I'd like to submit a PR with an updated option on the CLI and maven plugin to accept a custom signing service. This would remove the need to Fork the codebase to implement the custom signing service. My intitial thoughts are:

- Introduce a `storetype` called `SIGNING_SERVICE`
- Utilise `keystore` to specify the class name of the custom signing service class.

Would appreciate feedback on this approach (or if there's already a mechanism to achieve this that I've missed).

**Comment by ebourg** [comment]
I like the idea of supporting custom store types. I've recently added a service provider interface for the Signable implementations and I tried to do the same for the store types, but I wasn't satisfied with the result as it exposed too much of the internal API. I'm still pondering what design to adopt.

Please don't hesitate to send a PR, I can't promise I'll merge it as is but it'll provide a basis for discussion. I wonder if the implementation class could be specified as part of the storetype parameter, something like `SIGNING_SERVICE:com.example.CustomSigningService`. That would preserve the keystore parameter for its intended usage. This design will also probably require some changes to the SigningService interface, there is no uniform way to initialize the services since the implementations have different constructors.
